### PR TITLE
Acceleration for x86 compressed String hashCode with vectorization

### PR DESCRIPTION
**Pull request by msqb**
For obtaining the hashCode of a compressed String, each iteration of the SSE loop moves 4 bytes from memory to the low doubleword of the XMM register. The stride shift amount is set to 0 for handling bytes. These packed 1 byte integers are then zero extended with instruction pmovzxbd and the same procedure to obtain the hashCode as with decompressed Strings is used.

A vector conversion example has been added to show how this procedure works.

Signed-off-by: Mansoor Saqib <Mansoor.Saqib@ibm.com>

### CODE REVIEW

**Comment by msqb** [comment]
Requires upstream OMR pull request [Adds x86 instruction PMOVZXBD](https://github.com/eclipse/omr/pull/2246). @andrewcraik

**Comment by andrewcraik** [comment]
Some comment on the performance evaluation of this change is warranted since this is a performance optimization/improvement - I believe @msqb is working on distilling his results to be posted here shortly.

File: runtime/compiler/trj9/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -11940,32 +11990,51 @@ inlineStringHashCode(
...
 
       generateLabelInstruction(LABEL, node, SSELabel, cg);
-      generateRegMemInstruction(MOVSDRegMem,
-                                node,
-                                xmm2,
-                                generateX86MemoryReference(valueReg, indexReg, 1, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg),
-                                cg);
+
+      // move 8 bytes (4 2-byte characters) to xmm2. Stride shift = 1 (each element in valueReg is 2 bytes. Used with indexReg for correct traversal)
+      if(!isCompressed)
```
Comment (fjeremic) [inline runtime/compiler/trj9/x/codegen/J9TreeEvaluator.cpp]: This if/else has both paths. Would it be possible to turn this into a positive query and avoid the negation?

Comment (msqb) [reply runtime/compiler/trj9/x/codegen/J9TreeEvaluator.cpp]: @fjeremic consulting OMR's compiler/z/codegen's inlineStringHashCode as a reference, it had the same type of negation if/else structure so I used that format. However if you'd like I can change it to become a positive query.

https://github.com/eclipse/omr/blob/ec2c3bf81b14c6ee01f5e03e8a35a730682ebb6a/compiler/z/codegen/OMRTreeEvaluator.cpp#L17252

Comment (andrewcraik) [reply runtime/compiler/trj9/x/codegen/J9TreeEvaluator.cpp]: I think avoiding the negation makes sense for new code - there are certainly dark areas in the existing code worthy of clean up but if we make things a quantum better every time a change is made the overall quality of everything moves up over time.

Comment (msqb) [reply runtime/compiler/trj9/x/codegen/J9TreeEvaluator.cpp]: sure, I'll make the change

Comment (fjeremic) [reply runtime/compiler/trj9/x/codegen/J9TreeEvaluator.cpp]: > However if you'd like I can change it to become a positive query.

Definitely fix up both while you're in this area of code. Thanks!

**Comment by msqb** [comment]
@fjeremic replaced negations with positive query

**Comment by msqb** [comment]
Performance analysis was conducted using a microbenchmark using strings of lengths 3, 12, and 48 characters long composed of English lowercase letters. For each of these length variations, String's hashCode() method was repeatedly called on 10,000 unique strings with steps taken to prevent cached hashcodes from being used as part of the measurement.

Previously when enabling string compression, String's hashcode was constructed serially. Comparing the current serial string hashCode() implementation for compressed strings to the vectorized implementation in this changeset, all three string lengths showed a 60% performance improvement when running OpenJ9 OpenJDK9 with string compression enabled. OpenJ9 OpenJDK8 showed a similar performance improvement with string compression enabled.

The performance of hashCode() with string compression disabled remained unchanged for both OpenJ9 OpenJDK9 and OpenJ9 OpenJDK8.

OpenJ9 OpenJDK9 and OpenJ9 OpenJDK8 used for this analysis were compiled from:
OpenJ9   - d7cc8fd
OMR       - 9bc3259

**Comment by andrewcraik** [comment]
@msqb this seems to have been lost in the churn - could you please rebase forward and I can then launch the appropriate sanity builds to try and get this merged?

**Comment by andrewcraik** [comment]
@msqb could you please rebase this forward so we can get it tested?

**Comment by andrewcraik** [comment]
@0dvictor could you let me know your thoughts on this change?

**Comment by 0dvictor** [comment]
The code looks good to me. It is a straightforward extension of current hash code implementation of uncompressed strings.
FYI, https://github.com/eclipse/openj9/pull/1647 has a better implementation then current one, and it supports both compressed and uncompressed strings.

**Comment by andrewcraik** [comment]
@0dvictor so is your suggestion that this should be closed?

**Comment by andrewcraik** [comment]
Jenkins test sanity

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux jdk8,jdk9,jdk10

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux jdk8,jdk10

**Comment by msqb** [comment]
I've rebased forward. Thanks

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win

**Comment by andrewcraik** [comment]
@msqb so we've had a bit of build instability - can you rebase forward so we can finally try to clear the x86 sanity tests to get this in?

**Comment by msqb** [comment]
@andrewcraik I've rebased forward. Thanks

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win

**Comment by andrewcraik** [comment]
x86 Linux is completely clean failures on windows look to be related to #2129

**Pull request merged**